src/lib/core/components/visualizations/InputVariables.tsx

```diff
@@ -180,22 +181,27 @@ export function InputVariables(props: Props) {
   return (
     <div>
       <div className={classes.inputs}>
-        {inputs.map((input, index) => (
-          <div key={input.name} className={classes.input}>
-            <div className={classes.label}>{input.label}</div>
-            <VariableTreeDropdown
-              rootEntity={entities[0]}
-              disabledVariables={disabledVariablesByInputIndex[index]}
-              starredVariables={starredVariables}
-              toggleStarredVariable={toggleStarredVariable}
-              entityId={values[input.name]?.entityId}
-              variableId={values[input.name]?.variableId}
-              onChange={(variable) => {
-                handleChange(input.name, variable);
-              }}
-            />
-          </div>
-        ))}
+        {inputs.map((input, index) => {
+          const Wrapper = input.wrapper ?? (({ children }) => <>{children}</>);
+          return (
+            <Wrapper>
+              <div key={input.name} className={classes.input}>
+                <div className={classes.label}>{input.label}</div>
+                <VariableTreeDropdown
+                  rootEntity={entities[0]}
+                  disabledVariables={disabledVariablesByInputIndex[index]}
+                  starredVariables={starredVariables}
+                  toggleStarredVariable={toggleStarredVariable}
+                  entityId={values[input.name]?.entityId}
+                  variableId={values[input.name]?.variableId}
+                  onChange={(variable) => {
+                    handleChange(input.name, variable);
+                  }}
+                />
+              </div>
+            </Wrapper>
```

Comment:
If you adopt my suggestion of giving `InputSpec` an `InputComponent`, you would do something like this:

```suggestion
            <InputComponent
              key={input.name}
              rootEntity={entities[0]}
              disabledVariables={disabledVariablesByInputIndex[index]}
              starredVariables={starredVariables}
              toggleStarredVariable={toggleStarredVariable}
              entityId={values[input.name]?.entityId}
              variableId={values[input.name]?.variableId}
              onChange={(variable) => {
                handleChange(input.name, variable);
              }}
            />
```

Reply:
Thanks Jamie - I'm having a go at this now.

When `InputComponent !== VariableTreeDropdown`, how do I pass the extra props that the extended component needs?

e.g. when `InputComponent` is `VariableTreeDropdownWithToggle` I need to pass these extra props:
```
interface Props extends VariableTreeDropdownProps {
  /** the text label to show with the toggle switch */
  toggleLabel: string;
  /** the controlled state of the toggle switch */
  toggleState: boolean | undefined;
  /** handler for toggle switch state change */
  onToggleStateChange: (newState: boolean) => void;
  /** optional flag to disble the toggle switch */
  disableToggle?: boolean
}
```
It seems that maybe
```
interface InputSpec {
  name: string;
  label: string;
  component?: React.FunctionComponent;
}
```
should include a `componentExtraProps: any` but that gets nasty...

Reply:
I've got as far as I can.
I don't know if we need to get into Generics, but I'm finding it hard to type the props correctly.  The branch below compiles, but fails at runtime because of the `: any` typing and the lack of "extra props" passed to the "special" component.

```
interface InputSpec {
  name: string;
  label: string;
  component?: (props: any) => JSX.Element; // ? extends VariableTreeDropdownProps
}
```


https://github.com/VEuPathDB/web-eda/tree/missing-data-button-test1

Reply:
Update - might abort the wrapper approach!  (See Dave's comments below.)

Reply:
No worries - Dave's suggestion of adding a `hasMissingDataToggle`  should do the job. 😄 

For future reference, the `InputComponent` approach that I was suggesting wouldn't have involved generics - the `InputComponent` would always receive the props the default `VariableTreeDropdown` expects, and these props would always be passed in `InputVariables`. If the `InputComponent` needs additional "props," they could be passed implicitly through whatever `useCallback` was used to derive the `InputComponent`.

(You did something similar for your initial wrapper work - there, the only "explicit" prop you were passing to the `Wrapper` was its `children`, and for the missing data wrapper, you passed the additional `MissingDataToggleWrapperProps` implicitly through the `useCallback` which defined that wrapper.)

Reply:
Thanks Jamie - it's good to know I was barking up the right tree!